### ISSUE TITLE
显存泄露

### ISSUE THREAD

**Issue by ZLT-a-LL** [ultralytics/ultralytics#22950]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

使用yolo11最新集成的sam3模型进行视频文本追踪时，显存越用越多是怎么回事，在服务器上从5G一直往上涨，最后48G显存都不够用，即使打开了stream=True，本地也是同样的情况

### Environment

(yolo11) PS E:\A_worker\W_XM\yolo11_car_seg\yolo_sam3> yolo checks
Ultralytics 8.3.237  Python-3.12.12 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB)
Setup complete  (20 CPUs, 31.7 GB RAM, 754.2/953.9 GB disk)

OS                     Windows-11-10.0.26100-SP0
Environment            Windows
Python                 3.12.12
Install                pip
Path                   E:\Miniconda\envs\yolo11\Lib\site-packages\ultralytics
RAM                    31.71 GB
Disk                   754.2/953.9 GB
CPU                    12th Gen Intel Core(TM) i7-12700H
CPU count              20
GPU                    NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB
GPU count              1
CUDA                   12.8

numpy                   2.2.6>=1.23.0
matplotlib              3.10.7>=3.3.0
opencv-python           4.12.0.88>=4.6.0
pillow                  11.3.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.3>=1.4.1
torch                   2.9.1+cu128>=1.8.0
torch                   2.9.1+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.24.1+cu128>=0.9.0
psutil                  7.1.3>=5.8.0
polars                  1.36.0>=0.20.0
ultralytics-thop        2.0.18>=2.0.18
(yolo11) PS E:\A_worker\W_XM\yolo11_car_seg\yolo_sam3> 

### Minimal Reproducible Example

import cv2
from ultralytics.models.sam import SAM3VideoSemanticPredictor

# Initialize semantic video predictor
overrides = dict(conf=0.25,
                 task="segment",
                 mode="predict",
                 imgsz=640,
                 model=r"E:\A_worker\W_XM\yolo11_sam3\models\sam3\sam3.pt",
                 half=True,
                 vid_stride=10,)
predictor = SAM3VideoSemanticPredictor(overrides=overrides,
                                       bpe_path=r"E:\A_worker\W_XM\yolo11_sam3\assets\bpe_simple_vocab_16e6.txt.gz")

# Track concepts using text prompts
results = predictor(source=r"E:\A_worker\W_XM\yolo11_car_seg\test_img\flow\2.mp4",
                    text=["vehicle"],
                    stream=True,
                    save=True)

# Process results
for r in results:
    img = r.plot()  # 处理当前帧，生成带追踪框的图像
    cv2.imshow('yolo_sam3', img)

    # 关键修改：将参数从 0 改为 1。程序会等待1毫秒并自动继续。
    # 同时检测是否按下了 'q' 键，以便随时退出。
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cv2.destroyAllWindows()

# Alternative: Track with bounding box prompts
# results = predictor(
#     source="path/to/video.mp4",
#     bboxes=[[864, 383, 975, 620], [705, 229, 782, 402]],
#     labels=[1, 1],  # Positive labels
#     stream=True,
#     save=True,
# )

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ZLT-a-LL, thank you for your interest in Ultralytics 🚀! This is an automated response to help us triage your issue quickly—an Ultralytics engineer will also assist soon. In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report (e.g., the VRAM usage growing over time during SAM3 video text tracking), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You’ve already shared a great script—if possible, please also include the smallest possible video/sample and exact command/script you run so we can reproduce the VRAM growth behavior reliably on our side 🔍.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
这更像是“按设计增长”而不是典型泄露：SAM3 视频追踪会把每一帧的 `maskmem_features/maskmem_pos_enc` 等记忆写进 `inference_state` 的 `output_dict` 里用于后续传播与交互，所以视频越长显存会持续上升，`stream=True` 只是不缓存可视化/结果流，并不会清掉追踪记忆（可对照 [`SAM3VideoSemanticPredictor._tracker_update_memories()` 参考实现](https://docs.ultralytics.com/reference/models/sam/predict/)）。建议先用很短的视频（比如 100–300 帧）验证是否仍“无上限”上涨；如果需要跑超长视频，目前的规避方式是分段处理并在段与段之间重建 predictor（或至少 `predictor.inference_state.clear()` 后再跑下一段）来释放历史帧记忆，同时把复现时的完整显存曲线/完整 traceback（文本）贴出来便于我们确认是否还有额外的缓存未释放。

**Comment by ZLT-a-LL** [comment]
好的，谢谢，我尝试一下

**Comment by glenn-jocher** [comment]
可以的，试完后如果显存仍然“无上限”上涨，麻烦你贴一下每隔 N 帧的显存日志（这样我们能区分是追踪记忆增长还是额外缓存未释放），例如：

```python
import torch

for i, r in enumerate(results):
    if torch.cuda.is_available() and i % 50 == 0:
        print(i, "MB alloc=", torch.cuda.memory_allocated() / 1024**2,
                 "MB max=", torch.cuda.max_memory_allocated() / 1024**2)
```

另外如果你要跑超长视频，最稳的做法是先把视频切成小段（每段 30–120 秒都行），每段单独跑一次 predictor（参考 [SAM3 追踪记忆写入位置的实现说明](https://docs.ultralytics.com/reference/models/sam/predict/)），例如用 `ffmpeg` 分段：

```bash
ffmpeg -i input.mp4 -c copy -map 0 -f segment -segment_time 60 -reset_timestamps 1 out_%03d.mp4
```

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue is now merged in [PR #22969: “Prune SAM3 tracker memory to prevent VRAM leak”](https://github.com/ultralytics/ultralytics/pull/22969) by @Y-T-G, with contributions from @Laughing-q.

### What changed (and why it may fix the leak)
This PR aims to keep long-running video tracking/segmentation memory usage bounded by automatically pruning SAM “non-conditioning” frame memory that could otherwise grow without limit:

- Enables a new default flag `self.clear_non_cond_mem = True` to periodically clear old non-conditioning memory.
- Adds a helper `_prune_non_cond_memory(...)` and calls it after each frame output is stored during:
  - standard `inference(...)`
  - video `propagate_in_video(...)`
- Prunes outdated entries from:
  - `output_dict["non_cond_frame_outputs"]`
  - each object’s `output_dict_per_obj[*]["non_cond_frame_outputs"]`
- Uses the model’s temporal memory settings (`num_maskmem` and `memory_temporal_stride_for_eval`) to decide how many recent frames to retain.

If the VRAM/RAM growth was caused by cached non-conditioning outputs accumulating over long videos, these changes should prevent that unbounded buildup.

### How to test
If testing from `main` is OK, installing directly from source is the fastest way to verify:

```bash
pip install -U git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, waiting for the next official PyPI release is also an option if preferred.

### Feedback requested
It would be helpful to know whether memory stays stable now when running the same long-video scenario that previously reproduced the leak (including approximate video length, device, and peak VRAM if available).

Thanks 🙏 for reporting this—any additional details are welcome if the issue persists.
